### PR TITLE
Opentelemetry traces and heapsize memory analysis

### PR DESCRIPTION
**Pull request by pepeiborra**
This PR builds on the work of @mpardalos and supersedes #826 

A new facility to log open telemetry events and spans to the GHC event log is added. The following are logged:
* LSP events (all of them)
* Shake action spans

Additionally, when the flag `--ot-memory-profiling` is given, the following are logged periodically:
* Count of items in the ShakeExtras "state" (Metric is called "values map size_bytes")
* Aggregate size of items per key, with sharing, so that shared costs are not reported twice.

For the non LSP driver, the flag `--ot-memory-profiling` produces a textual version of the above, instead of an eventlog.

There is documentation on how to visualise the eventlogs using Tracy. The Nix script takes care of installing all the dependencies

TODOs:

- [x] Merge my heapsize changes in https://github.com/mpardalos/Heapsize/pull/1, release a new version to Hackage and remove the git submodule

Future TODOs (not for this PR):
* Use spans to link LSP events and Shake actions, as well as Shake dependencies
* `--ot-memory-profiling` should be always enabled and log only on SIGHUP(or some other signal) instead of periodically
* Output mechanism for traces should be configurable, so that we can log to stdout, the cloud, etc.

### CODE REVIEW

**Comment by pepeiborra** [comment]
Some examples of memory analysis for the ghcide project itself:

```
cabal run ghcide src/Development/IDE.hs -- --ot-memory-profiling  -j2 +RTS -A4G -Sstderr 2> /dev/null
...
Completed (1 file worked, 0 files failed)
# Shake value store contents(976):
  - GhcSession: 231036.42KB
  - GhcSessionDeps: 1.18KB
  - GetModificationTime_ {missingFileDiagnostics = False}: 6.65KB
  - GetModSummaryWithoutTimestamps: 761.74KB
  - GetModIfaceWithoutLinkable: 7.39KB
  - IsHiFileStable: 7.12KB
  - GetLocatedImports: 658.78KB
  - GetFileExists: 15.47KB
  - GetFileContents: 8.98KB
  - NeedsCompilation: 7.19KB
  - GetModIface: 632.13KB
  - GetModSummary: 7.07KB
  - IsFileOfInterest: 7.26KB
  - ReportImportCycles: 7.21KB
  - GetDependencyInformation: 463.82KB
  - GetModIfaceFromDisk: 6.94KB
  - GetDependencies: 140.36KB
  - GenerateCore: 263.22KB
  - GetKnownTargets: 16.95KB
  - GetHieAst: 25.62KB
  - GetModuleGraph: 28.58KB
  - GetParsedModule: 35.09KB
  - GetFilesOfInterest: 2.22KB
  - TypeCheck: 65.98KB
  - GhcSessionIO: 11101.30KB
Total: 245.31MB
```
```
cabal run ghcide src/Development/IDE.hs -- --ot-memory-profiling  -j2 +RTS -A4G -Sstderr 2> /dev/null
...
Completed (50 files worked, 0 files failed)
# Shake value store contents(1119):
  - GhcSession: 295383.43KB
  - GhcSessionDeps: 2035.29KB
  - GetModificationTime_ {missingFileDiagnostics = False}: 4.85KB
  - GetModSummaryWithoutTimestamps: 760.94KB
  - GetModIfaceWithoutLinkable: 7.50KB
  - GetLocatedImports: 660.46KB
  - GenerateCore: 91166.96KB
  - GetFileExists: 15.39KB
  - TypeCheck: 27222.14KB
  - GetFileContents: 8.98KB
  - NeedsCompilation: 7.19KB
  - GetModIface: 342.98KB
  - GetHieAst: 15515.02KB
  - GetModSummary: 7.07KB
  - IsFileOfInterest: 8.04KB
  - ReportImportCycles: 7.21KB
  - GetDependencyInformation: 456.74KB
  - GetDependencies: 140.36KB
  - GetParsedModule: 107.39KB
  - GetModIfaceFromDisk: 0.27KB
  - IsHiFileStable: 3.28KB
  - GetKnownTargets: 109.21KB
  - GetModuleGraph: 28.58KB
  - GetFilesOfInterest: 104.14KB
  - GhcSessionIO: 11375.06KB
Total: 445.48MB
```

**Comment by pepeiborra** [comment]
@wz1000 `GenerateCore` is crazy expensive 😜

**Comment by wz1000** [comment]
I don't understand why running the same command twice results in such different output.

**Comment by wz1000** [comment]
> @wz1000 GenerateCore is crazy expensive

In terms of memory, time or both?

File: src/Development/IDE/Core/Tracing.hs
Comment (wz1000) [inline src/Development/IDE/Core/Tracing.hs:162]: Maybe we want a sensible upper bound for the recursion here, so it doesn't repeat forever.

File: src/Development/IDE/Core/Tracing.hs
Comment (wz1000) [inline src/Development/IDE/Core/Tracing.hs:101]: Do we need a `performGC` here?

**Comment by pepeiborra** [comment]
> I don't understand why running the same command twice results in such different output.

It's not the same command, copy and Paste mistake

File: src/Development/IDE/Core/Tracing.hs
Comment (wz1000) [inline src/Development/IDE/Core/Tracing.hs:144]: I think GC can't run while there is a `unsafe` FFI call is in progress, so maybe we can fork a thread that calls a C function which waits for the analysis to complete (as signaled by a write to the file system or something).

**Comment by pepeiborra** [comment]
> > @wz1000 GenerateCore is crazy expensive
> 
> 
> 
> In terms of memory, time or both?

Memory. It's several times more expensive than parsing or typechecking

**Comment by wz1000** [comment]
> Memory. It's several times more expensive than parsing or typechecking

Hmm. I don't think we need it for anything other than warnings and generating bytecode when needed, so maybe we can avoid keeping it in the graph.

File: src/Development/IDE/Core/Tracing.hs
Comment (pepeiborra) [reply src/Development/IDE/Core/Tracing.hs:144]: This is correct, but the Haskell threads do not run either. 

The only way imho to take advantage of this property is to rewrite the heap size analysis in C

File: src/Development/IDE/Core/Tracing.hs
Comment (pepeiborra) [reply src/Development/IDE/Core/Tracing.hs:162]: Yes, indeed

File: src/Development/IDE/Core/Tracing.hs
Comment (pepeiborra) [reply src/Development/IDE/Core/Tracing.hs:101]: `runHeapsize` takes care of that

**Comment by wz1000** [comment]
I'm also wondering about the interaction of `heapsize` with the concurrent garbage collector. Will it work?

**Comment by pepeiborra** [comment]
> I'm also wondering about the interaction of `heapsize` with the concurrent garbage collector. Will it still work?

What do you mean by concurrent garbage collector? If you mean `--non-moving` then it should work better by definition.

File: src/Development/IDE/Core/Tracing.hs
Comment (wz1000) [reply src/Development/IDE/Core/Tracing.hs:174]: Andreas tells me on #ghc that this is not true, and Haskell threads will continue to run as long we are using the threaded runtime.

Comment (pepeiborra) [reply src/Development/IDE/Core/Tracing.hs:174]: I think that's probably a misunderstanding between you and Andreas, but I'm happy to be corrected by a program that demonstrates your theory. 

In the meantime, I'll quote https://downloads.haskell.org/ghc/latest/docs/html/users_guide/runtime_control.html#rts-flag---long-gc-sync=<seconds>

**Comment by wz1000** [comment]
If I understand correctly, using `heapsize` to measure is inherently unreliable, since if we run out of nursery while the measurement is running, the GC has to run before we can proceed.

To work around this, we need to do the entire measurement in C. Fortunately, IOHK seems to have already written something that does this: https://github.com/input-output-hk/cardano-prelude/blob/df335a2dc1505a379052732dc3e8cec094330bc9/cardano-prelude/cbits/closure_size.c

@bgamari can you confirm if my summary is correct.

**Comment by pepeiborra** [comment]
Correct, yes. 

What we have now will suffice for smallish codebases, but will not scale to medium ones. The trick to detect GC using a weak ref seems to be working pretty well in Linux, and as long as the nursery is big enough, the heapsize analysis is stable. It's been running for 2 hours on the ghcide codebase without a hiccup, collecting 24 samples with an `-A4G` nursery.

**Comment by pepeiborra** [comment]
The time statistics are also useful. `GetHieAst` is pretty expensive... 

image

**Comment by pepeiborra** [comment]
@jneira help please, the stack windows build is upset and I have no idea why

**Comment by jneira** [comment]
@pepeiborra i would try to remove `-f-external-interpreter`, we never have been sure about its effectiveness.
If build fails without it, a expeditive option would be switch stack.windows.yaml to ghc-8.6.5, the unique version reliable enough in wondows for now

**Comment by pepeiborra** [comment]
> @pepeiborra i would try to remove `-f-external-interpreter`, we never have been sure about its effectiveness.
> If build fails without it, a expeditive option would be switch stack.windows.yaml to ghc-8.6.5, the unique version reliable enough in windows for now

Dropping the `-fexternal-interpreter` flag didn't help. I doubt switching to ghc-8.6.5 will help either, since the problem here is not reliability but an actual stack build issue. The error suggests that the heapsize package descriptor is probably missing some library paths.

To help unblock this, could you please build this branch locally both with Cabal and Stack, and compare the output of `ghc-pkg describe heapsize`?

**Review by wz1000 — APPROVED** [review]

**Comment by jneira** [comment]
Yeah, the error is the same with and without using an external interpreter, will try asap

**Comment by pepeiborra** [comment]
@jneira any news?

**Comment by jneira** [comment]
After building the project with satck and the default stack.yaml (8.10.2) which throws the same error and `cabal -w ghc-8.10.2` :
```
D:\ws\haskell\ghcide>ghc-pkg --user-package-db .\dist-newstyle\packagedb\ghc-8.10.2 describe heapsize
name:                 heapsize
version:              0.2
visibility:           public
id:                   heapsize-0.2-inplace
key:                  heapsize-0.2-inplace
license:              BSD-3-Clause
copyright:            Michail Pardalos 2020
maintainer:           Michail Pardalos <mpardalos@gmail.com>
author:               Michail Pardalos <mpardalos@gmail.com>
synopsis:             Determine the size of runtime data structures
description:
    heapsize is a tool to determine the size data structures.
    Determining the size of recursive data structures is
    supported. All sizes are in Bytes.

category:             GHC, Debug, Development
abi:                  inplace
exposed:              True
exposed-modules:      HeapSize
import-dirs:
    D:\ws\haskell\ghcide\dist-newstyle\build\x86_64-windows\ghc-8.10.2\heapsize-0.2\build

library-dirs:
    D:\ws\haskell\ghcide\dist-newstyle\build\x86_64-windows\ghc-8.10.2\heapsize-0.2\build
    D:\\OpenSSL-x86_64\lib

dynamic-library-dirs:
    D:\ws\haskell\ghcide\dist-newstyle\build\x86_64-windows\ghc-8.10.2\heapsize-0.2\build
    D:\\OpenSSL-x86_64\lib

data-dir:             D:\ws\haskell\ghcide\heapsize
hs-libraries:         HSheapsize-0.2-inplace
include-dirs:         D:\\OpenSSL-x86_64\include
depends:
    base-4.14.1.0 deepseq-1.4.4.0 exceptions-0.10.4 ghc-heap-8.10.2
    hashable-1.3.0.0-7eef361fa2b3a306ca7c6d89c695f4465e42392f
    hashtables-1.2.4.1-5aae77ea53a5219107f0f91948fe2ddc070b4fce
    primitive-0.7.1.0-e605ac3dff285b52b30a1168db6f30b40298dcfa
    transformers-0.5.6.2
    unordered-con_-0.2.13.0-de9d6aea4ef6b26e708f2f25368f02bdd7b0822a

haddock-interfaces:
    D:\ws\haskell\ghcide\dist-newstyle\build\x86_64-windows\ghc-8.10.2\heapsize-0.2\doc\html\heapsize\heapsize.haddock

haddock-html:
    D:\ws\haskell\ghcide\dist-newstyle\build\x86_64-windows\ghc-8.10.2\heapsize-0.2\doc\html\heapsize
pkgroot: "D:\\ws\\haskell\\ghcide\\dist-newstyle\\packagedb"


D:\ws\haskell\ghcide>stack exec ghc-pkg -- describe heapsize
name:                 heapsize
version:              0.2
visibility:           public
id:                   heapsize-0.2-9mPl8uIsnRT9ry8OwVZpm8
key:                  heapsize-0.2-9mPl8uIsnRT9ry8OwVZpm8
license:              BSD-3-Clause
copyright:            Michail Pardalos 2020
maintainer:           Michail Pardalos <mpardalos@gmail.com>
author:               Michail Pardalos <mpardalos@gmail.com>
synopsis:             Determine the size of runtime data structures
description:
    heapsize is a tool to determine the size data structures.
    Determining the size of recursive data structures is
    supported. All sizes are in Bytes.

category:             GHC, Debug, Development
abi:                  8f28cb17ebb384607e86e25840020e68
exposed:              True
exposed-modules:      HeapSize
import-dirs:
    D:\ws\haskell\ghcide\.stack-work\install\63c8a773\lib\x86_64-windows-ghc-8.10.2\heapsize-0.2-9mPl8uIsnRT9ry8OwVZpm8

library-dirs:
    D:\ws\haskell\ghcide\.stack-work\install\63c8a773\lib\x86_64-windows-ghc-8.10.2\heapsize-0.2-9mPl8uIsnRT9ry8OwVZpm8
    D:\\bin\Programs\stack\x86_64-windows\msys2-20180531\mingw64\lib
    D:\\bin\Programs\stack\x86_64-windows\msys2-20180531\mingw64\bin

dynamic-library-dirs:
    D:\ws\haskell\ghcide\.stack-work\install\63c8a773\lib\x86_64-windows-ghc-8.10.2
    D:\\bin\Programs\stack\x86_64-windows\msys2-20180531\mingw64\lib
    D:\\bin\Programs\stack\x86_64-windows\msys2-20180531\mingw64\bin

data-dir:
    D:\ws\haskell\ghcide\.stack-work\install\63c8a773\share\x86_64-windows-ghc-8.10.2\heapsize-0.2

hs-libraries:         HSheapsize-0.2-9mPl8uIsnRT9ry8OwVZpm8
include-dirs:
    D:\\bin\Programs\stack\x86_64-windows\msys2-20180531\mingw64\include

depends:
    base-4.14.1.0 deepseq-1.4.4.0 exceptions-0.10.4 ghc-heap-8.10.2
    hashable-1.3.0.0-1P2Y3eA5cTdB8Gcn7yS7qi
    hashtables-1.2.3.4-FUeV0APkAq1Jayqid8tkiB
    primitive-0.7.1.0-Jxsyd70oUttYiCXCa0HqV transformers-0.5.6.2
    unordered-containers-0.2.12.0-BkJSZe2LvIB4fc0ASSWCag

haddock-interfaces:
    D:\ws\haskell\ghcide\.stack-work\install\63c8a773\doc\heapsize-0.2\heapsize.haddock

haddock-html:
    D:\ws\haskell\ghcide\.stack-work\install\63c8a773\doc\heapsize-0.2
pkgroot: "D:\\ws\\haskell\\ghcide\\.stack-work\\install\\63c8a773"

````

**Comment by pepeiborra** [comment]
Thanks @jneira. What's the output of `ghc-pkg describe ghc-heap` in each case?

**Comment by jneira** [comment]
@pepeiborra I've separated them in the original comment

**Comment by pepeiborra** [comment]
Sorry I wasn't clear. The output of `ghc-pkg describe heapsize` didn't show anything relevant, so now I am asking for the output of `ghc-pkg describe ghc-heap`. I guess I could also ask CI for this instead of bugging you.

**Comment by jneira** [comment]
Oh sorry i misreaded ghc-heap. Dont worry, but warm stack cache takes some time and i am in other machine now and i have to build from zero, i'll post it

**Comment by pepeiborra** [comment]
@jneira don't worry, it's faster for me to ask CI

**Comment by mpardalos** [comment]
@pepeiborra, just wanted to say thanks for taking this over from #826.
Hopefully I can address some of your TODOs from this PR once my university workload dies down a bit.

**Comment by pepeiborra** [comment]
@jneira I changed the windows-stack CI script to build the dependencies first, then `stack test`, and now it's failing in a very mysterious way that I cannot understand. Could you take a look please?

**Comment by jneira** [comment]
@pepeiborra will try to take a look asap, but the error does not look good, we could disable temporaly the windows build if you want to go forward and activate it afterwards. 
In hls there is no stack build for windows so will no suffer it.

File: stack-windows.yaml
Comment (jneira) [inline stack-windows.yaml:58]: OMG :facepalm: 

Thanks for the detailed report of the root cause

Comment (pepeiborra) [reply stack-windows.yaml:58]: Yeah, but the fix is not enough. There is another stack issue after it.

**Comment by pepeiborra** [comment]
@jneira @ndmitchell unless one of you can get to the bottom of the problem here, I will drop the Stack Windows build as we lack the maintainer power to keep it working.

**Comment by jneira** [comment]
The error seems to be of the same kind:
```
ghcide> ghc.exe:  | D:\a\1\s\.stack-work\dist\a3a5fe88\build\HSghcide-0.5.0-GfeKdUCOB448C8Yqgu22oa.o: unknown symbol `heapsizzezm0zi3zi0zmG6oW1FfbuDgJS3obQNcUGF_HeapSizze_runHeapsizze1_closure'
```

**Comment by pepeiborra** [comment]
https://github.com/haskell/ghcide/pull/934 opened to unblock this PR

**Pull request merged**
